src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,225 @@
+            }
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+            if (changedSpanCollection.IsEmpty())
+            {
+                return;
+            }
+
+            var viewLines = TextView.TextViewLines;
+            using var _ = PooledDictionary<int, IMappingTagSpan<InlineDiagnosticsTag>>.GetInstance(out var map);
+            GetSpansOnEachLine(changedSpanCollection, map);
+            foreach (var (lineNum, tagMappingSpan) in map)
+            {
+                // Mapping the IMappingTagSpan back up to the TextView's visual snapshot to ensure there will
+                // be no adornments drawn on disjoint spans.
+                if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span))
+                {
+                    continue;
+                }
+
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry is null)
+                {
+                    continue;
+                }
+
+                // Need to get the SnapshotPoint to be able to get the IWpfTextViewLine
+                var point = tagMappingSpan.Span.Start.GetPoint(TextView.TextSnapshot, PositionAffinity.Predecessor);
+                if (point == null)
+                {
+                    continue;
+                }
+
+                var tag = tagMappingSpan.Tag;
+                var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.GetClassificationId(tag.ErrorType));
+                var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+
+                var lineView = TextView.GetTextViewLineContainingBufferPosition(point.Value);
+
+                var visualElement = graphicsResult.VisualElement;
+                Canvas.SetLeft(visualElement,
+                    tag.Location == InlineDiagnosticsLocations.PlacedAtEndOfCode ? lineView.Right :
+                    tag.Location == InlineDiagnosticsLocations.PlacedAtEndOfEditor ? TextView.ViewportRight - visualElement.DesiredSize.Width :
+                    throw ExceptionUtilities.UnexpectedValue(tag.Location));
+
+                Canvas.SetTop(visualElement, geometry.Bounds.Bottom - visualElement.DesiredSize.Height);
+
+                // Only place the diagnostics if the diagnostic would not intersect with the editor window
+                if (lineView.Right < TextView.ViewportWidth - visualElement.DesiredSize.Width)
```

Comment:
so this seems to be a critical check you shoudl do prior to any other unnecessary work.  othereise this method read weird: do all this prep, then throw the results away.